### PR TITLE
feature #4065 - gas and gas price can be changed for dapp transactions

### LINKED ISSUE #4065: It should be possible to change gas and gasprise for transactions from dapps or extensions
status-go https://github.com/status-im/status-go/issues/881

### PR DESCRIPTION
**Pull request by goranjovic**
fixes #4065 

### Summary:

Adds the option to pass gas and gas limit for transactions created in dapps. The changes involve:

- `status-go` upgrade to `develop-g68966af0` for the dependency endpoints
- updated ios and android native modules to use the new endpoints
- updated UI to allow gas editing for dapp transactions

### Testing notes (optional):

Every send transaction option, not just from dapp (e.g. Wallet Send, chat command, both ether and tokens, etc.) - all where touched by the change.

### Steps to test:
- Send a transaction from dapp
- Tap Advanced, open gas settings and change them
- Send transaction and check the gas details in Transaction history and on Etherscan
- The same thing for other transaction flows

status: ready

### CODE REVIEW

**Review by janherich — APPROVED** [review]

File: src/status_im/native_module/core.cljs [hunk 1]
```diff
@@ -40,8 +40,11 @@
 (defn login [address password callback]
   (module-interface/-login rns-module address password callback))
 
-(defn approve-sign-requests [hashes password callback]
-  (module-interface/-approve-sign-requests rns-module hashes password callback))
+(defn approve-sign-request [id password callback]
+  (module-interface/-approve-sign-request rns-module id password callback))
+
+(defn approve-sign-request-with-args [id password gas gas-price callback]
```
Comment (flexsurfer) [inline src/status_im/native_module/core.cljs:46]: why do we need two methods? maybe we could just set gas and gas-price `nil` and do not use `approve-sign-request` ?

File: src/status_im/ui/screens/wallet/send/views.cljs [hunk 2]
```diff
@@ -106,6 +106,11 @@
   (when (and gas gas-price)
     (money/wei->ether (.times gas gas-price))))
 
+(defn return-to-transaction [modal?]
+  (if modal?
+    (re-frame/dispatch [:navigate-to-modal :wallet-send-transaction-modal])
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/views.cljs]: custom navigation? something wrong, it shouldn't be possible to open modal from modal, we should discuss this with UX

Comment (goranjovic) [reply src/status_im/ui/screens/wallet/send/views.cljs]: There's a good reason, let's discuss in chat.

File: src/status_im/native_module/core.cljs (hunk 1 above)
Comment (goranjovic) [reply src/status_im/native_module/core.cljs:46]: It's a constraint on go side and I prefer to keep 1:1 between go endpoints and native module functions.

**Review by mandrigin — APPROVED** [review]
LGTM

**Review by rasom — APPROVED** [review]

**Comment by lukaszfryc** [comment]
For some reason which maybe unrelated to this PR, I can't send messages from iOS build:


Everything works fine on Android (this PR build) or latest nightly (`03243e-n-fl`). @goranjovic is there any chance that this PR is related to this?

**Comment by lukaszfryc** [comment]
Btw, I checked transactions in dapp (deploying a contract and getting STT in Sample DApp). That worked fine. I could my gas settings in transaction history and on etherscan. 

I still need to check wallet and chat transactions.

**Comment by lukaszfryc** [comment]
@goranjovic one thing I've already noticed that we should fix is: Clicking on the "x" in transaction fee screen for a dapp closes the whole transaction instead of going back to the previous screen:



Steps:
1. Open Sample Dapp
2. Click deploy contract to sign the transaction
3. Open transaction fee screen
4. Click "x" to go back to the previous screen
-> You end up in the dapp instead of sign transaction screen

**Comment by lukaszfryc** [comment]
@goranjovic I've tested the latest build, here is what I found:

1) I still have the "x" issue described in the previous comment.

2) Also, I see that default gas price and limit are not shown in transaction fee screen. 


What's more, when I set these values manually and click done button, the sign transaction screen still shows the old values


3) If I set too low gas in a dapp (for instance Airswap) I get the "intrinsic gas to low" error which is OK. But, when I go to the transaction history, I'm getting a BigNumber error:

**Comment by goranjovic** [comment]
@lukaszfryc Ok, but is chat still broken?

I was unsuccessfully trying to reproduce the transaction history error (3) over the weekend for another bugfix. This is mainnet right?

**Comment by lukaszfryc** [comment]
> Ok, but is chat still broken?

No, all transactions should work fine now.

Yes, it's mainnet. I'm able to reproduce it 2/2 for Airswap

**Comment by goranjovic** [comment]
@lukaszfryc Still can't reproduce the transaction history, but see a different problem. Are you able to Sign messages? I'm getting a number format exception (both Simple dapp on Ropsten and Airswap on Mainnet)

**Comment by lukaszfryc** [comment]
@goranjovic I was able to sign them when I tested it couple of hours ago.

**Comment by lukaszfryc** [comment]
@goranjovic re transaction history, let me try it again on a new build.

**Comment by goranjovic** [comment]
@lukaszfryc re re transaction history, what keeps happening to me is that transaction with low gas never even appears in history, but no errors. Just to clarify, are you setting a low gas limit, gas price or both?

**Comment by goranjovic** [comment]
@lukaszfryc Found the cause for the crash I was seeing on Sign message. Really weird it doesn't happen for you. Fixed it and will push along with close with X as soon as I address the empty gas values issue.

**Comment by goranjovic** [comment]
@lukaszfryc Ready to be re-tested. I addressed all points except (3) `If I set too low gas in a dapp...`  which I still can't reproduce. 
That issue I'm addressing in #4712  so if you notice anything around that please comment there.

**Comment by lukaszfryc** [comment]
@goranjovic 1, 2 work fine but there are 2 different issues now that occur only on Android:

1) Android status bar turns white when Transaction Fee for signing a dapp transaction is shown: 

It looks fine in chat transactions:


2) I now can reproduce the format exception when trying to request token or sign a contract in Sample Dapp. The error shows after providing a password and clicking "Sign transaction". Everything works fine for other dapps like Airswap or for chat transactions.   
```
{"timestamp":1528793123,"message":"java.lang.NumberFormatException: null","stackTrace":"java.lang.NumberFormatException: null\n\tat java.lang.Long.parseLong(Long.java:406)\n\tat java.lang.Long.parseLong(Long.java:485)\n\tat im.status.ethereum.module.StatusModule$9.run(StatusModule.java:535)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)\n\tat java.lang.Thread.run(Thread.java:762)\n"}
```

**Comment by goranjovic** [comment]
@lukaszfryc Fixed and pushed. Good catch for status bar, it was actually broken in Chat Send as well, just wasn't visible because the screen behind it was also blue.

**Comment by lukaszfryc** [comment]
@goranjovic works great 👍But, can you fix "the [BigNumber error](https://github.com/status-im/status-react/pull/4682#issuecomment-396173830) in transaction history" in scope of this PR? Otherwise we're going to introduce this issue in `develop`.

**Comment by goranjovic** [comment]
@lukaszfryc That issue wasn't introduced in this PR and I'd rather not expand the scope too much, this is a really big PR as it is. And besides, that bug has its own PR already.

**Comment by lukaszfryc** [comment]
@goranjovic I'm only worried that by allowing users to set custom gas in dapps, they may encounter this issue more often. I could not reproduce it with wallet or chat transactions in recent nightly.

**Comment by goranjovic** [comment]
@lukaszfryc And I could not reproduce it at all. When I try it the transaction just stays at 0 confirmations, which is expected - if the gas price is too low, nobody will want to mine it, fair enough. But I never see the error in transaction history.
Seeing the same behavior for chat/wallet transactions in the other PR and here.
What are the exact steps you do when you get this? Maybe I'm trying something slightly different.

**Comment by lukaszfryc** [comment]
@goranjovic I can't reproduce it now either. Previously when it happened I set low gas and got the intrinsict gas too low. Then when I went to the transaction history I got the error popup. Anyway, I think we can create a separate issue for it.

But please give me ~20 more min, I still want to check few things.

**Pull request merged**